src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -0,0 +1,92 @@
+        /// Note that DiagnosticHandler is the latest in the pipeline </param>
+        public DiagnosticsHandler(HttpMessageHandler innerHandler) : base(innerHandler)
+        {
+        }
+
+        internal static bool IsEnabled()
+        {
+            return s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ResponseWriteName);
+        }
+
+        protected internal override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,
+            CancellationToken cancellationToken)
+        {
+            //HttpClientHandler is responsible to call DelegatingHandler.IsEnabled() before forwarding request here.
+            //This code will not be reached if ResponseWriteName is not enabled, unless consumer unsubscribes
+            //from DiagnosticListener right after the check. So some requests happening right after subscription starts
+            //might not be instrumented. Similarly, when consumer unsubscribes, extra requests might be instumented
+
+            Guid loggingRequestId = Guid.NewGuid();
+            if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.RequestWriteName))
+            {
+                LogHttpRequest(request,  loggingRequestId);
+            }
+            Task<HttpResponseMessage> responseTask = base.SendAsync(request, cancellationToken);
+
+            LogHttpResponse(responseTask, loggingRequestId);
+            return responseTask;
+        }
+
+        #region private
+
+        private static readonly DiagnosticListener s_diagnosticListener =
+            new DiagnosticListener(DiagnosticsHandlerLoggingStrings.DiagnosticListenerName);
+
+        private static void LogHttpRequest(HttpRequestMessage request, Guid loggingRequestId)
+        {
+            long timestamp = Stopwatch.GetTimestamp();
+
+            s_diagnosticListener.Write(
+                DiagnosticsHandlerLoggingStrings.RequestWriteName,
+                new
+                {
+                    Request = request,
+                    LoggingRequestId = loggingRequestId,
+                    Timestamp = timestamp
+                }
+            );
+        }
+
+        private static void LogHttpResponse(Task<HttpResponseMessage> responseTask, Guid loggingRequestId)
+        {
+            responseTask.ContinueWith(
+                (t, s) =>
+                {
+                    long timestamp = Stopwatch.GetTimestamp();
+
+                    s_diagnosticListener.Write(
+                        DiagnosticsHandlerLoggingStrings.ResponseWriteName,
+                        new
+                        {
+                            Response = t.Status == TaskStatus.RanToCompletion ? t.Result : null,
+                            LoggingRequestId = (Guid)s,
+                            TimeStamp = timestamp,
+                            Exception = t.Exception,
+                            RequestTaskStatus = t.Status
+                        }
```

Comment:
#13172 looks to "Send failed HTTP requests to DiagnosticSource"... does one determine an exception case by looking to see if `.Exception` is null or not? Not that this has to follow what we do else where but most other events that I can think of have a separate event for exceptions and try not to "overload" the exception case into the end event.